### PR TITLE
Implement bot skill and listbots

### PR DESCRIPTION
**Pull request by DolceTriade**

### CODE REVIEW

File: src/sgame/sg_bot_public.h [hunk 1]
```diff
@@ -69,6 +70,7 @@ void G_BotCleanup();
 void G_BotFill( bool immediately );
 void G_BotRemoveObstacle( qhandle_t handle );
 void G_BotUpdateObstacles();
+std::string btos( gentity_t *bot );
```
Comment (necessarily-equal) [inline src/sgame/sg_bot_public.h:73]: I huh, am not sure having a global function with a undescriptive name is a good idea. Even if `vtos` or `etos` do exist

File: src/sgame/sg_bot.cpp [hunk 2]
```diff
@@ -690,3 +703,35 @@ void G_SetBotFill( int fill )
...
+
+std::string btos( gentity_t *bot )
+{
+	if ( !( bot->r.svFlags & SVF_BOT ) )
+	{
+		return "";
+	}
+	return Str::Format( "^*%s^*: %s [s=%d b=%d g=%s]",
+			bot->client->pers.netname, BG_TeamName( G_Team( bot ) ), bot->botMind->botSkill.level,
+			bot->botMind->behaviorTree->name, BotGoalToString( bot ).c_str() );
```
Comment (necessarily-equal) [inline src/sgame/sg_bot.cpp:736]: I don't think you need the `c_str()`
```suggestion
			bot->botMind->behaviorTree->name, BotGoalToString( bot ) );
```

**Review by necessarily-equal — CHANGES_REQUESTED** [review]
Having a way to set bot skill sounds good. Some notes:

File: src/sgame/sg_admin.cpp [hunk 3]
```diff
@@ -5999,6 +6000,32 @@ bool G_admin_bot( gentity_t *ent )
...
+		int skill = BotSkillFromString( ent, skillStr.c_str() );
+		team_t team = TEAM_NONE;
+		if ( args.Argc() > 3 )
+		{
+			const std::string& teamStr = args.Argv( 3 );
+			team = BG_PlayableTeamFromString( teamStr.c_str() );
+		}
+		for ( int i = 0; i < MAX_CLIENTS; ++i )
+		{
+			if ( team != TEAM_NONE && G_Team( &g_entities[ i ] ) != team )
```
Comment (slipher) [inline src/sgame/sg_admin.cpp:6028]: I think this is missing an `ent->inuse` check. Consider using ForEntities<ClientComponent>

Comment (necessarily-equal) [reply src/sgame/sg_admin.cpp:6028]: or `for ( gentity_t *ent : iterate_client_entities )` :eyes:

File: src/sgame/sg_bot.cpp (hunk 2 above)
Comment (DolceTriade) [reply src/sgame/sg_bot.cpp:736]: Correct, thanks.

File: src/sgame/sg_admin.cpp (hunk 3 above)
Comment (DolceTriade) [reply src/sgame/sg_admin.cpp:6028]: Done.

File: src/sgame/sg_bot_public.h (hunk 1 above)
Comment (DolceTriade) [reply src/sgame/sg_bot_public.h]: Renamed.

**Comment by necessarily-equal** [comment]
LGTM, haven't tested

**Review by necessarily-equal — APPROVED** [review]

**Comment by necessarily-equal** [comment]
I could test this a bit, it seems like it works well.


For testing convenience, this is deployed as a testing server, although rebased a bit by me, and with a commit that shouldn't affect this PR.

> Experimental Development Server - fix-defaultFill" / https://play.unvanquished.net/172.86.178.34:27992

File: src/sgame/sg_bot.cpp [hunk 4]
```diff
@@ -690,3 +703,35 @@ void G_SetBotFill( int fill )
...
+	return "<unknown goal>";
+}
+
+std::string G_BotToString( gentity_t *bot )
+{
+	if ( !( bot->r.svFlags & SVF_BOT ) )
+	{
+		return "";
+	}
+	return Str::Format( "^*%s^*: %s [s=%d b=%d g=%s]",
```
Comment (slipher) [inline src/sgame/sg_bot.cpp:734]: `b=%d` is using %d for a string. Otherwise LGTM

**Comment by necessarily-equal** [comment]
gentle bump

**Comment by DolceTriade** [comment]
Sorry, fix the comments and merge later tonight.

**Pull request merged**